Chaining view::transform 3 times with one transform2 fails to compile with gcc

Hello,

I've encontered an issue when trying to compile a simple use case with gcc 8.2. I have the following code:

```cpp
std::vector<int> buff(12, -1);
::ranges::span<int> sp(buff.data(), 12);

auto x = ::ranges::view::transform(sp, [](int a) { return a > 3 ? a : 42; });
auto y = ::ranges::view::transform(x, sp, [](int a, int b) { return a + b; });
auto rng = ::ranges::view::transform(y, [](int a) { return a + 1; });
```

The last line fails to compile with gcc 8.2 and issue the following error:
```bash
/.../.conan/data/range-v3/0.4.0/ericniebler/stable/package/e52aec0ecc24318bcb1daf3dc03e2f3e38ea6e23/include/range/v3/view/transform.hpp:280:73: error:
could not convert ‘{((const ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >*)this)->ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >::fun_, ranges::v3::CPOs::begin.ranges::v3::_begin_::fn::operator()<const ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> > >(((const ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >*)this)->ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >::rng1_), ranges::v3::CPOs::begin.ranges::v3::_begin_::fn::operator()<const ranges::v3::span<int> >(((const ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >*)this)->ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >::rng2_)}’
from ‘<brace-enclosed initializer list>’
to
‘ranges::v3::iter_transform2_view<ranges::v3::transform_view<ranges::v3::span<int>, Range_range_transform2_read_chain_Test::TestBody()::<lambda(int)> >, ranges::v3::span<int>, ranges::v3::indirected<Range_range_transform2_read_chain_Test::TestBody()::<lambda(int, int)> > >::cursor’
                 return {fun_, ranges::begin(rng1_), ranges::begin(rng2_)};
```
(full stack in attachment)
[error_full.txt](https://github.com/ericniebler/range-v3/files/2778973/error_full.txt)

When I've investigated a little bit further, I tried to change the code of view::transform:
```cpp
CONCEPT_REQUIRES(Range<Rng1 const>() && Range<Rng2 const>())
cursor begin_cursor() const
{
    return {fun_, ranges::begin(rng1_), ranges::begin(rng2_)};
}
```

by:
```cpp
CONCEPT_REQUIRES(Range<Rng1 const>() && Range<Rng2 const>())
cursor begin_cursor() const
{
    iterator_t<Rng2> it2 = ranges::begin(rng2_);
    iterator_t<Rng1> it1 = ranges::begin(rng1_);
    return {fun_, it1, it2};
}
```

Now I get this error:
```cpp
/.../.conan/data/range-v3/0.4.0/ericniebler/stable/package/e52aec0ecc24318bcb1daf3dc03e2f3e38ea6e23/include/range/v3/view/transform.hpp:278:53:
error: conversion from
‘basic_iterator<adaptor_cursor<[...],adaptor<true>>>’
to non-scalar type
‘basic_iterator<adaptor_cursor<[...],adaptor<false>>>’ requested
                 iterator_t<Rng1> it1 = ranges::begin(rng1_);
                                        ~~~~~~~~~~~~~^~~~~~~
```
(full stack in attachment)
[error_full2.txt](https://github.com/ericniebler/range-v3/files/2778967/error_full2.txt)

After investigating the conversion constructors, it seems that `basic_iterator` explicitly have one, adaptor too, but I didn't see one for `adaptor_cursor` (except for a `using base_t::base_t`, but maybe it is not even needed for `adaptor_cursor` since there are no members). As for why the "const" boolean is true in one case and false in the other, I don't know (bad deduction/overload somewhere ?)

Clang7 seems to compile it so I don't know if it is a compiler bug or a library bug.

You can find a live repro on [godbolt](https://godbolt.org/z/eP8fKZ).

--- a/include/range/v3/view/transform.hpp
+++ b/include/range/v3/view/transform.hpp
@@ -126,12 +126,12 @@ namespace ranges
             CONCEPT_REQUIRES(SizedRange<Rng const>())
             constexpr range_size_type_t<Rng> size() const
             {
-                return ranges::size(this->base());
+                return static_cast<range_size_type_t<Rng>>(ranges::size(this->base()));
             }
             CONCEPT_REQUIRES(SizedRange<Rng>())
             RANGES_CXX14_CONSTEXPR range_size_type_t<Rng> size()
             {
-                return ranges::size(this->base());
+                return static_cast<range_size_type_t<Rng>>(ranges::size(this->base()));
             }
         };
 

--- a/include/range/v3/view/transform.hpp
+++ b/include/range/v3/view/transform.hpp
@@ -81,7 +81,7 @@ namespace ranges
                   : fun_(std::move(fun))
                 {}
                 template<bool Other,
-                    CONCEPT_REQUIRES_(IsConst && !Other)>
+                    CONCEPT_REQUIRES_(IsConst && (!Other))>
                 adaptor(adaptor<Other> that)
                   : fun_(std::move(that.fun_))
                 {}
@@ -166,41 +166,62 @@ namespace ranges
                 range_cardinality<Rng1>::value,
                 range_cardinality<Rng2>::value);
 
+            template<bool>
             struct cursor;
 
+            template<bool Const>
             struct sentinel
             {
             private:
-                friend struct cursor;
-                sentinel_t<Rng1> end1_;
-                sentinel_t<Rng2> end2_;
+                friend struct cursor<Const>;
+                sentinel_t<meta::const_if_c<Const, Rng1>> end1_;
+                sentinel_t<meta::const_if_c<Const, Rng2>> end2_;
             public:
                 sentinel() = default;
-                sentinel(detail::any, sentinel_t<Rng1> end1, sentinel_t<Rng2> end2)
-                  : end1_(std::move(end1)), end2_(std::move(end2))
+                sentinel(meta::const_if_c<Const, iter_transform2_view> &parent, decltype(end))
+                  : end1_(end(parent.rng1_)), end2_(end(parent.rng2_))
+                {}
+                template<bool Other, CONCEPT_REQUIRES_(Const && (!Other))>
+                sentinel(sentinel<Other> that)
+                  : end1_(std::move(that.end1_))
+                  , end2_(std::move(that.end2_))
                 {}
             };
 
+            template<bool Const>
             struct cursor
             {
             private:
-                using fun_ref_ = semiregular_ref_or_val_t<Fun, true>;
+                using fun_ref_ = semiregular_ref_or_val_t<Fun, Const>;
+                using R1 = meta::const_if_c<Const, Rng1>;
+                using R2 = meta::const_if_c<Const, Rng2>;
                 fun_ref_ fun_;
-                iterator_t<Rng1> it1_;
-                iterator_t<Rng2> it2_;
+                iterator_t<R1> it1_;
+                iterator_t<R2> it2_;
 
             public:
                 using difference_type = difference_type_;
                 using single_pass = meta::or_c<
-                    (bool) SinglePass<iterator_t<Rng1>>(),
-                    (bool) SinglePass<iterator_t<Rng2>>()>;
+                    (bool) SinglePass<iterator_t<R1>>(),
+                    (bool) SinglePass<iterator_t<R2>>()>;
                 using value_type =
-                    detail::decay_t<invoke_result_t<Fun &, copy_tag, iterator_t<Rng1>,
-                        iterator_t<Rng2>>>;
+                    detail::decay_t<
+                        invoke_result_t<
+                            meta::const_if_c<Const, Fun> &,
+                            copy_tag,
+                            iterator_t<R1>,
+                            iterator_t<R2>>>;
 
                 cursor() = default;
-                cursor(fun_ref_ fun, iterator_t<Rng1> it1, iterator_t<Rng2> it2)
-                  : fun_(std::move(fun)), it1_(std::move(it1)), it2_(std::move(it2))
+                template<typename BeginEndFn>
+                cursor(meta::const_if_c<Const, iter_transform2_view> &parent, BeginEndFn begin_end)
+                  : fun_(parent.fun_), it1_(begin_end(parent.rng1_)), it2_(begin_end(parent.rng2_))
+                {}
+                template<bool Other, CONCEPT_REQUIRES_(Const && (!Other))>
+                cursor(cursor<Other> that)
+                  : fun_(std::move(that.fun_))
+                  , it1_(std::move(that.end1_))
+                  , it2_(std::move(that.end2_))
                 {}
                 auto read() const
                 RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
@@ -219,28 +240,28 @@ namespace ranges
                     // one reaches the end.
                     return it1_ == that.it1_ || it2_ == that.it2_;
                 }
-                bool equal(sentinel const &s) const
+                bool equal(sentinel<Const> const &s) const
                 {
                     // By returning true if *any* of the iterators are equal, we allow
                     // transformed ranges to be of different lengths, stopping when the first
                     // one reaches the end.
                     return it1_ == s.end1_ || it2_ == s.end2_;
                 }
-                CONCEPT_REQUIRES(BidirectionalRange<Rng1>() && BidirectionalRange<Rng2>())
+                CONCEPT_REQUIRES(BidirectionalRange<R1>() && BidirectionalRange<R2>())
                 void prev()
                 {
                     --it1_;
                     --it2_;
                 }
-                CONCEPT_REQUIRES(RandomAccessRange<Rng1>() && RandomAccessRange<Rng2>())
+                CONCEPT_REQUIRES(RandomAccessRange<R1>() && RandomAccessRange<R2>())
                 void advance(difference_type n)
                 {
                     ranges::advance(it1_, n);
                     ranges::advance(it2_, n);
                 }
                 CONCEPT_REQUIRES(
-                    SizedSentinel<iterator_t<Rng1>, iterator_t<Rng1>>() &&
-                    SizedSentinel<iterator_t<Rng2>, iterator_t<Rng2>>())
+                    SizedSentinel<iterator_t<R1>, iterator_t<R2>>() &&
+                    SizedSentinel<iterator_t<R1>, iterator_t<R2>>())
                 difference_type distance_to(cursor const &that) const
                 {
                     // Return the smallest distance (in magnitude) of any of the iterator
@@ -255,31 +276,38 @@ namespace ranges
                 )
             };
 
+            template<bool Const>
             using end_cursor_t =
                 meta::if_c<
-                    BoundedRange<Rng1>() && BoundedRange<Rng2>() &&
-                        !SinglePass<iterator_t<Rng1>>() &&
-                        !SinglePass<iterator_t<Rng2>>(),
-                    cursor,
-                    sentinel>;
+                    BoundedRange<meta::const_if_c<Const, Rng1>>() &&
+                    BoundedRange<meta::const_if_c<Const, Rng2>>() &&
+                    !SinglePass<iterator_t<meta::const_if_c<Const, Rng1>>>() &&
+                    !SinglePass<iterator_t<meta::const_if_c<Const, Rng2>>>(),
+                    cursor<Const>,
+                    sentinel<Const>>;
 
-            cursor begin_cursor()
+            cursor<simple_view<Rng1>() && simple_view<Rng2>()>
+            begin_cursor()
             {
-                return {fun_, ranges::begin(rng1_), ranges::begin(rng2_)};
+                return {*this, ranges::begin};
             }
-            end_cursor_t end_cursor()
+            end_cursor_t<simple_view<Rng1>() && simple_view<Rng2>()> end_cursor()
             {
-                return {fun_, ranges::end(rng1_), ranges::end(rng2_)};
+                return {*this, ranges::end};
             }
-            CONCEPT_REQUIRES(Range<Rng1 const>() && Range<Rng2 const>())
-            cursor begin_cursor() const
+            template<typename CRng1 = Rng1 const, typename CRng2 = Rng2 const,
+                CONCEPT_REQUIRES_(Range<CRng1>() && Range<CRng2>() &&
+                    IndirectInvocable<Fun const &, iterator_t<CRng1>, iterator_t<CRng2>>())>
+            cursor<true> begin_cursor() const
             {
-                return {fun_, ranges::begin(rng1_), ranges::begin(rng2_)};
+                return {*this, ranges::begin};
             }
-            CONCEPT_REQUIRES(Range<Rng1 const>() && Range<Rng2 const>())
-            end_cursor_t end_cursor() const
+            template<bool Const = true, typename CRng1 = Rng1 const, typename CRng2 = Rng2 const,
+                CONCEPT_REQUIRES_(Range<CRng1>() && Range<CRng2>() &&
+                    IndirectInvocable<Fun const &, iterator_t<CRng1>, iterator_t<CRng2>>())>
+            end_cursor_t<Const> end_cursor() const
             {
-                return {fun_, ranges::end(rng1_), ranges::end(rng2_)};
+                return {*this, ranges::end};
             }
             template<class Self>
             static constexpr size_type_ size_(Self& self)

--- a/test/view/transform.cpp
+++ b/test/view/transform.cpp
@@ -17,6 +17,7 @@
 #include <range/v3/view/transform.hpp>
 #include <range/v3/view/counted.hpp>
 #include <range/v3/view/reverse.hpp>
+#include <range/v3/span.hpp>
 #include <range/v3/view/zip.hpp>
 #include <range/v3/algorithm/move.hpp>
 #include <range/v3/utility/copy.hpp>
@@ -30,6 +31,18 @@ struct is_odd
         return (i % 2) == 1;
     }
 };
+
+// https://github.com/ericniebler/range-v3/issues/996
+void bug_996()
+{
+    std::vector<int> buff(12, -1);
+    ::ranges::span<int> sp(buff.data(), 12);
+
+    auto x = ::ranges::view::transform(sp, [](int a) { return a > 3 ? a : 42; });
+    auto y = ::ranges::view::transform(x, sp, [](int a, int b) { return a + b; });
+    auto rng = ::ranges::view::transform(y, [](int a) { return a + 1; });
+    (void)rng;
+}
 
 int main()
 {

